### DISCUSSION TITLE
Automating async relationship loading in FastAPI response with AsyncAttrs (SQLAlchemy/SQLModel)

### DISCUSSION THREAD

**Discussion by DurandA** [fastapi/fastapi#10986] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import List, Type, TypeVar
from contextlib import asynccontextmanager

from fastapi import Depends, FastAPI, HTTPException, Query
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.orm import sessionmaker, selectinload
from sqlalchemy.ext.asyncio import AsyncAttrs
from sqlalchemy.orm import RelationshipProperty
from sqlalchemy.future import select
from sqlalchemy.inspection import inspect
from sqlmodel import (Field, Relationship, SQLModel,
                      select)
from sqlmodel.ext.asyncio.session import AsyncSession


class Parent(SQLModel, AsyncAttrs, table=True):
    id: int | None = Field(default=None, primary_key=True, index=True, nullable=False)
    name: str
    children: list["Child"] = Relationship(back_populates="parent")


class Child(SQLModel, AsyncAttrs, table=True):
    id: int | None = Field(default=None, primary_key=True, index=True, nullable=False)
    name: str
    parent_id: int = Field(default=None, foreign_key="parent.id")
    parent: Parent | None = Relationship(back_populates="children")


class ChildRead(SQLModel):
    id: int
    name: str


class ParentRead(SQLModel):
    id: int
    name: str
    children: list[Parent]


T = TypeVar('T', bound=SQLModel)

async def load_relationships_and_create_pydantic(
    db_obj, 
    pydantic_model: Type[T], 
    session: AsyncSession
) -> T:
    # create a dictionary to hold the attributes including relationships
    loaded_attrs = {}

    for field_name, field_info in pydantic_model.__annotations__.items():
        mapper_attrs = inspect(db_obj.__class__).mapper.attrs
        # check if the field is a relationship that should be loaded
        if field_name in mapper_attrs and isinstance(mapper_attrs[field_name], RelationshipProperty):
            relationship_data = await getattr(db_obj.awaitable_attrs, field_name)
            if isinstance(relationship_data, list): # relationship is a list of items, use type of the list
                loaded_attrs[field_name] = [
                    await load_relationships_and_create_pydantic(item, field_info.__args__[0], session)
                    for item in relationship_data
                ]
            else: # relationship is a single item
                loaded_attrs[field_name] = await load_relationships_and_create_pydantic(relationship_data, field_info, session)
        else:
            # regular attribute, get its value
            attr_value = getattr(db_obj, field_name, None)
            if attr_value is not None:
                loaded_attrs[field_name] = attr_value

    # create the Pydantic instance
    return pydantic_model(**loaded_attrs)


sqlite_file_name = "database.db"
sqlite_url = f"sqlite+aiosqlite:///{sqlite_file_name}"

connect_args = {"check_same_thread": False}
engine = create_async_engine(sqlite_url, echo=True, connect_args=connect_args)


async def create_db_and_tables():
    async with engine.begin() as conn:
        await conn.run_sync(SQLModel.metadata.create_all)


AsyncSessionLocal = sessionmaker(
    bind=engine, class_=AsyncSession, expire_on_commit=False
)


async def get_session() -> AsyncSession:
    async with AsyncSessionLocal() as session:
        yield session


@asynccontextmanager
async def lifespan(app: FastAPI):
    await create_db_and_tables()
    yield


app = FastAPI(lifespan=lifespan)


@app.get("/parents/{parent_id}", response_model=ParentRead)
async def read_parent(*, session: AsyncSession = Depends(get_session), parent_id: int):
    db_parent = await session.get(Parent, parent_id)
    if not db_parent:
        raise HTTPException(status_code=404, detail="Parent not found")
    parent = await load_relationships_and_create_pydantic(db_parent, ParentRead, session)
    return parent
```

### Description

I want to streamline the process of loading relationships asynchronously according to the `response_model` when a SQLModel or SQLAlchemy object is returned by a route function. For more context, this is when an object is loaded from a database using an asynchronous driver but its relationships are not yet loaded.

SQLAlchemy 2.0 has an [`AsyncAttrs` mixin](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs) which enables loading relationships/attributes using [`AsyncAttrs.awaitable_attrs`](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs.awaitable_attrs).

I developped a proof of concept mainly consisting of the `load_relationships_and_create_pydantic()` function which:
1. takes a SQLModel instance, inspects its fields, and if a field that is a relationship is encountered, load this relationship asynchronously.
2. for each related object, recursively calls itself, ensuring that even nested relationships are loaded
3. converts these loaded models and their relationships into Pydantic models suitable for response serialization

However, it does not integrate very well with FastAPI as `load_relationships_and_create_pydantic()` has to be called with the right parameters at the end of each route.

Is there a way to specify a custom validator that is called before calling [`create_response_field()`](https://github.com/tiangolo/fastapi/blob/d74b3b25659b42233a669f032529880de8bd6c2d/fastapi/routing.py#L484-L488)?

### Operating System

Linux

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

Python 3.12.1

### Additional Context

It is probably not possible to do exactly what I want without forking FastAPI. I am however looking for advices and maybe some general guidelines if I would choose to fork FastAPI with the long term goal to make a PR.

**Comment by bogdan-coman-uv** [comment, marked as answer]
I don't think you need all that. You can define your relationships to be eagerly loaded (with lazy="selectin") and the just return the db model. It should be converted automatically to a pydantic model.

For example:

```
from sqlalchemy.ext.declarative import declarative_base
BaseSql = declarative_base()

from sqlalchemy.orm import relationship
import sqlalchemy as sa

class Parent(BaseSql):
     __tablename__ = "parents"
    id: int  = sa.Column(sa.Integer(), primary_key=True)
    name: str = sa.Column(sa.String(255))
    children: "list[Child]" = relationship(back_populates="parent", lazy="selectin")
  
class Child(BaseSql):
   __tablename__ = "children"
    id: int = sa.Column(sa.Integer(), primary_key=True)
    name: str = sa.Column(sa.String(255))
    parent_id: int|None = sa.Column(sa.Integer(), sa.ForeignKey("parents.id"))
    parent: Parent|None = relationship(back_populates="children")

from pydantic import BaseModel
class ORMBaseModel(BaseModel):
    model_config = ConfigDict(from_attributes=True) # https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.from_attributes

class ChildRead(ORMBaseModel):
    id:int
    name:str|None = None
    parent_id:int|None = None

class ParentRead(ORMBaseModel):
   id:int
   name:str|None = None
   children:list[ChildRead]

...

@app.get("/parents/{parent_id}", response_model=ParentRead)
async def read_parent(*, session: AsyncSession = Depends(get_session), parent_id: int):
    db_parent = await session.get(Parent, parent_id)
    if not db_parent:
        raise HTTPException(status_code=404, detail="Parent not found")
    
    return db_parent
```

**Reply by moloch--** [reply]
This has not been the case in my experience, relationships are not included in the response JSON even if they're populated in the server-side object using `selectin`.
